### ISSUE TITLE
RET503 false positive on exhaustive type check

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ruff#18656]
### Summary

RET503 still reports an error after an exhaustive type check. If the fix is applied, type checkers will warn/error about unreachable code.
[playground link](https://play.ruff.rs/98b1a534-fca2-4f66-859c-6be9a207f73f)

```py
import tarfile
import zipfile

type ArchiveKind = tarfile.TarFile | zipfile.ZipFile

def get_first_archive_member(archive: ArchiveKind) -> str:
    if isinstance(archive, tarfile.TarFile):
        return archive.getnames()[0]
    elif isinstance(archive, zipfile.ZipFile):
        return archive.namelist()[0]
    # RET503 error
```

### Version

ruff 0.11.13 (5faf72a4d 2025-06-05) + playground

**Comment by MichaReiser** [comment]
I can see how it would be desired that Ruff wouldn't raise `RET503` here. However, this requires that Ruff understands that `ArchiveKind` only has two variants and they're all handled. This is far beyond what Ruff's semantic model can handle today, but should be possible once we start using the semantic model from our in progress type checker ty in Ruff. 

For now, the best you can do here is to suppress `RET503`
